### ISSUE TITLE
Not all ipv6 port forwardings are created

### ISSUE THREAD

**Issue by rpasing** [moby/moby#45459]
### Description

On a clean install of Arch Linux using experimental ip6tables support of Docker, I have the problem that not all ports specified in my docker-compose.yml are actually open by the correct ip6tables rules.

I use the following docker-compose.yml as a quick way for reproducing the issue:
```
version: '2.4'

services:
  nginx:
    image: nginx
    ports:
      - 80:80
      - 443:443

networks:
   default:
     driver: bridge
     enable_ipv6: true
     ipam:
       config:
         - subnet: "fdcb:c299:00a6:87b2::/64"
```

Starting it via `docker-compose up -d` leads to the following:

`iptables-save | grep 80`:
```
-A POSTROUTING -s 172.19.0.2/32 -d 172.19.0.2/32 -p tcp -m tcp --dport 80 -j MASQUERADE
-A DOCKER ! -i br-c410199451e5 -p tcp -m tcp --dport 80 -j DNAT --to-destination 172.19.0.2:80
-A DOCKER -d 172.19.0.2/32 ! -i br-c410199451e5 -o br-c410199451e5 -p tcp -m tcp --dport 80 -j ACCEPT
```
`iptables-save | grep 443`:
```
-A POSTROUTING -s 172.19.0.2/32 -d 172.19.0.2/32 -p tcp -m tcp --dport 443 -j MASQUERADE
-A DOCKER ! -i br-c410199451e5 -p tcp -m tcp --dport 443 -j DNAT --to-destination 172.19.0.2:443
-A DOCKER -d 172.19.0.2/32 ! -i br-c410199451e5 -o br-c410199451e5 -p tcp -m tcp --dport 443 -j ACCEPT
```

`ip6tables-save | grep 80`:
```
-A DOCKER ! -i br-c410199451e5 -p tcp -m tcp --dport 80 -j DNAT --to-destination [fdcb:c299:a6:87b2::2]:80
```
`ip6tables-save | grep 443`:
```
-A POSTROUTING -s fdcb:c299:a6:87b2::2/128 -d fdcb:c299:a6:87b2::2/128 -p tcp -m tcp --dport 443 -j MASQUERADE
-A DOCKER ! -i br-c410199451e5 -p tcp -m tcp --dport 443 -j DNAT --to-destination [fdcb:c299:a6:87b2::2]:443
-A DOCKER -d fdcb:c299:a6:87b2::2/128 ! -i br-c410199451e5 -o br-c410199451e5 -p tcp -m tcp --dport 443 -j ACCEPT
```

As you see, for port 80 the postrouting and the forwarding rule is missing. This leads to that this container is not reachable via ipv6 on port 80 anymore (packets get lost due to forwarding is incomplete).

I see the following in the journal, line 1 seems like adding the route to the network is happening before the link is actually up:
```
May 03 22:17:11 archlinux dockerd[398]: time="2023-05-03T22:17:11.404752970Z" level=error msg="Could not add route to IPv6 network fdcb:c299:a6:87b2::1/64 via device br-f08902a4764e: network is down"
May 03 22:17:11 archlinux systemd-networkd[170]: br-f08902a4764e: Link UP
May 03 22:17:11 archlinux dockerd[398]: time="2023-05-03T22:17:11.509890277Z" level=error msg="Error handling plugin refcount operation" driver= error="legacy plugin: Unable to find plugin without name" mode=1
May 03 22:17:11 archlinux systemd[1]: var-lib-docker-overlay2-6eaf4ac8f0ef45411aa2e8b761fbdfe26120c6fc7ea06382724248d3ada7d73c\x2dinit-merged.mount: Deactivated successfully.
May 03 22:17:11 archlinux systemd[1]: var-lib-docker-overlay2-6eaf4ac8f0ef45411aa2e8b761fbdfe26120c6fc7ea06382724248d3ada7d73c-merged.mount: Deactivated successfully.
May 03 22:17:11 archlinux kernel: br-f08902a4764e: port 1(veth8ba7be4) entered blocking state
May 03 22:17:11 archlinux kernel: br-f08902a4764e: port 1(veth8ba7be4) entered disabled state
May 03 22:17:11 archlinux kernel: device veth8ba7be4 entered promiscuous mode
May 03 22:17:11 archlinux systemd-networkd[170]: veth8ba7be4: Link UP
May 03 22:17:11 archlinux kernel: br-f08902a4764e: port 1(veth8ba7be4) entered blocking state
May 03 22:17:11 archlinux kernel: br-f08902a4764e: port 1(veth8ba7be4) entered forwarding state
May 03 22:17:11 archlinux kernel: br-f08902a4764e: port 1(veth8ba7be4) entered disabled state
May 03 22:17:11 archlinux dockerd[398]: time="2023-05-03T22:17:11.584760320Z" level=info msg="No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [nameserver 8.8.8.8 nameserver 8.8.4.4]"
May 03 22:17:11 archlinux dockerd[398]: time="2023-05-03T22:17:11.584791907Z" level=info msg="IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844]"
```

I tried to reproduce on a cleanly installed Debian machine, but there I could not reproduce it. Seems related to using Arch Linux. I tried to use older versions of Docker (20.x), Containerd (1.6), Linux (LTS 6.1), but the issue still is there.

Maybe I'm doing something wrong. I'd be glad to get some help to try and find the cause for this. Thanks.

### Reproduce

1. Install a clean Arch Linux
2. `pacman -S docker docker-compose`
3. Paste below contents of `/etc/docker/daemon.json` into that file
4. `systemctl enable --now docker`
5. Paste above contents of ` docker-compose.yml`  into that file
6. `docker-compose up -d`
7. Compare ` iptables-save`  and `ip6tables-save`; port 80 is missing

### Expected behavior

Port 80 should be forwarded in the same manner for ipv4 and ipv6.

### docker version

```bash
Client:
 Version:           23.0.5
 API version:       1.42
 Go version:        go1.20.4
 Git commit:        bc4487a59e
 Built:             Tue May  2 21:53:09 2023
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          23.0.5
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       94d3ad69cc
  Built:            Tue May  2 21:53:09 2023
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.7.0
  GitCommit:        1fbd70374134b891f97ce19c70b6e50c7b9f4e0d.m
 runc:
  Version:          1.1.7
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc.)
    Version:  2.17.3
    Path:     /usr/lib/docker/cli-plugins/docker-compose

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 23.0.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1fbd70374134b891f97ce19c70b6e50c7b9f4e0d.m
 runc version:
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.2.13-arch1-1
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 1
 Total Memory: 1.879GiB
 Name: archlinux
 ID: b74122f2-b949-46f3-8ad7-b81841e4a897
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

/etc/docker/daemon.json:
```
{
    "ipv6": true,
    "fixed-cidr-v6": "fddb:c186:0b84:cc0a::/64",
    "experimental": true,
    "ip6tables": true
}
```

**Comment by tuxmainy** [comment]
Hi,
confirming this bug. I was running traefik with multiple port bindings & IPv6 and I noticed that only a few ports were available from external networks. The missing rules are all ip6tables forwarding rules in the format of:

```bash
ip6tables -A DOCKER ! -i $ipv6_bridge_iface -o $ipv6_bridge_iface --destination $ipv6_container_ip --proto $port_proto --dport $port_no --jump ACCEPT
```

As a workaround I wrote a shell script for firing up my container. If one is interested:
**ATTENTION! There this script has no error handling.**
```bash
#!/bin/bash

echo "Starting Containers"
docker compose up -d

NETWORK='your_ipv6_network'
CONTAINER='your_container-using-your_ipv6_network'

network_id=`docker network inspect $NETWORK |jq '.[].Id' |xargs echo`
ipv6_bridge_iface="br-${network_id:0:12}"

ipv6_container_ip=`docker container inspect $CONTAINER |jq ".[].NetworkSettings.Networks.$NETWORK.GlobalIPv6Address" |xargs echo`

ports=`docker container inspect internal-traefik-server-1 |jq '.[].NetworkSettings.Ports | keys[]' |xargs echo`

echo "Fixing ip6tables rules ..."
for port in $ports; do
	port_no=`echo $port |cut -d '/' -f 1`
	port_proto=`echo $port |cut -d '/' -f 2`

	echo -e "\t... $port"

	# clean rules which were successfully created by docker
	ip6tables -D DOCKER ! -i $ipv6_bridge_iface -o $ipv6_bridge_iface --destination $ipv6_container_ip --proto $port_proto --dport $port_no --jump ACCEPT 2>/dev/null

	ip6tables -A DOCKER ! -i $ipv6_bridge_iface -o $ipv6_bridge_iface --destination $ipv6_container_ip --proto $port_proto --dport $port_no --jump ACCEPT
done
```
### docker version
```
Client:
 Version:           23.0.5-ce
 API version:       1.42
 Go version:        go1.19.9
 Git commit:        94d3ad69cc59
 Built:             Thu Apr 27 00:00:00 2023
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          23.0.5-ce
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.9
  Git commit:       94d3ad69cc59
  Built:            Thu Apr 27 00:00:00 2023
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.6.20
  GitCommit:        2806fc1057397dbaeefbea0e4e17bddfbd388f38
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061b76bb
 docker-init:
  Version:          0.1.7_catatonit
  GitCommit:        
```
### docker info
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc.)
    Version:  2.18.1
    Path:     /usr/lib/docker/cli-plugins/docker-compose

Server:
 Containers: 36
  Running: 35
  Paused: 0
  Stopped: 1
 Images: 126
 Server Version: 23.0.5-ce
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: oci runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2806fc1057397dbaeefbea0e4e17bddfbd388f38
 runc version: v1.1.7-0-g860f061b76bb
 init version: 
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.3.2-1-default
 Operating System: openSUSE Tumbleweed
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 62.58GiB
 Name: devloop
 ID: AB4I:RA5K:EY26:FRJI:GAXR:D3IH:XTF7:FSVQ:JU5I:FMPZ:K5Y7:WYTA
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

regards
Daniel

**Comment by akerouanton** [comment]
Hi @tuxmainy, can you try to [turn on debug logs](https://docs.docker.com/config/daemon/logs/#enable-debugging) and paste here all logs related to iptables please?

**Comment by tuxmainy** [comment]
> Hi @tuxmainy, can you try to [turn on debug logs](https://docs.docker.com/config/daemon/logs/#enable-debugging) and paste here all logs related to iptables please?

Hi, yes, I'll do. But it will take a few days.

regards
Daniel

**Comment by tuxmainy** [comment]
> Hi @tuxmainy, can you try to [turn on debug logs](https://docs.docker.com/config/daemon/logs/#enable-debugging) and paste here all logs related to iptables please?

Wow. That is strange. In the executions you can clearly see the e.g. the commands to add NAT for port 8022. But it is not present in the chain O_o

What was executed according to the logs:
```
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.349560770+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C DOCKER -p tcp -d <my public IPv6 address> --dport 10051 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:10051 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.355375148+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A DOCKER -p tcp -d <my public IPv6 address> --dport 10051 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:10051 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.360668225+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 10051 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.365886485+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -A DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 10051 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.371236434+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C POSTROUTING -p tcp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 10051 -j MASQUERADE]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.376434112+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A POSTROUTING -p tcp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 10051 -j MASQUERADE]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.429514608+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C DOCKER -p tcp -d <my public IPv6 address> --dport 8022 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:8022 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.435076315+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A DOCKER -p tcp -d <my public IPv6 address> --dport 8022 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:8022 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.440459420+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 8022 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.445594104+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C POSTROUTING -p tcp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 8022 -j MASQUERADE]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.497296350+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C DOCKER -p tcp -d <my public IPv6 address> --dport 2222 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:2222 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.502811995+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A DOCKER -p tcp -d <my public IPv6 address> --dport 2222 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:2222 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.508061099+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 2222 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.511785862+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -A DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 2222 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.514898168+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C POSTROUTING -p tcp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 2222 -j MASQUERADE]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.535691852+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C DOCKER -p tcp -d <my public IPv6 address> --dport 443 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:443 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.537701596+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A DOCKER -p tcp -d <my public IPv6 address> --dport 443 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:443 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.539695819+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 443 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.541563763+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -A DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 443 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.543464446+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C POSTROUTING -p tcp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 443 -j MASQUERADE]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.563320968+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C DOCKER -p tcp -d <my public IPv6 address> --dport 80 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:80 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.565249319+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A DOCKER -p tcp -d <my public IPv6 address> --dport 80 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:80 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.567146982+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p tcp -d fd00:c0fe:babe:100::3 --dport 80 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.569009610+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C POSTROUTING -p tcp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 80 -j MASQUERADE]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.605064516+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C DOCKER -p udp -d <my public IPv6 address> --dport 53 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:53 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.607492293+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A DOCKER -p udp -d <my public IPv6 address> --dport 53 -j DNAT --to-destination [fd00:c0fe:babe:100::3]:53 ! -i br-a5a1d72a0261]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.609730091+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -C DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p udp -d fd00:c0fe:babe:100::3 --dport 53 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.611977183+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t filter -A DOCKER ! -i br-a5a1d72a0261 -o br-a5a1d72a0261 -p udp -d fd00:c0fe:babe:100::3 --dport 53 -j ACCEPT]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.614274875+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -C POSTROUTING -p udp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 53 -j MASQUERADE]"
Jun 28 20:32:07 devloop dockerd[23400]: time="2023-06-28T20:32:07.616567279+02:00" level=debug msg="/usr/sbin/ip6tables, [--wait -t nat -A POSTROUTING -p udp -s fd00:c0fe:babe:100::3 -d fd00:c0fe:babe:100::3 --dport 53 -j MASQUERADE]"
```

How the ip6tables DOCKER chain actually looked like
```
Chain DOCKER (3 references)
 pkts bytes target     prot opt in     out     source               destination         
   12   960 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:10051
    0     0 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:2222
    0     0 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:443
    4   412 ACCEPT     17   --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  udp dpt:53
```

How the ip6tables DOCKER chain should have looked like
```
Chain DOCKER (3 references)
 pkts bytes target     prot opt in     out     source               destination         
    2   160 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:10051
    0     0 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:2222
    0     0 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:443
    0     0 ACCEPT     17   --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  udp dpt:53
    0     0 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:80
    0     0 ACCEPT     6    --  !br-a5a1d72a0261 br-a5a1d72a0261  ::/0                 fd00:c0fe:babe:100::3  tcp dpt:8022
```

**Comment by zoredache** [comment]
I think I am running into the same issue.  I am starting to suspect it is related to a bug in ip6tables.  Specifically the `-C` option seems broken on `ip6tables v1.8.9`, but was not broken on `ip6tables v1.8.7`.

I have a gist where I show the output of a few ip6tables commands showing a different output between the two versions of iptables.

- https://gist.github.com/zoredache/08f571b9712ac8501a38649e6fcaae3d

**Comment by tuxmainy** [comment]
> So the hypothesis in the first post "seems like adding the route to the network is happening before the link is actually up" sounds plausible , because if its dependent on timing that would explain why sometimes it works and sometimes it don't.

I don't think so. If you look at my log you see that the iptables commands of port 10051 are executed before port 8022. But the resulting table is having 10051 but not 8022. If it would be a matter of "link up" the order should have been the other way round.

**Comment by tuxmainy** [comment]
> Yesterday a new version of iptables (v1.8.10) was released which includes the patch @zoredache mentioned.

Is [this](https://git.netfilter.org/iptables/commit/?id=78850e7dba64a949c440dbdbe557f59409c6db48) the fix we are talking about? Would make sense because I just noticed that only the last rule having the same amount of digits in the port can be seen after the container start. So one could think that the prior rules were removed by following. And the issue is something about "delete_mask" without understanding what exactly goes on...
